src/controls.js

```diff
@@ -201,63 +201,308 @@ Crafty.extend({
+                        }
+                    this.fingers[idx].entity = closest;
+                    this.fingers[idx].realX = x;
+                    this.fingers[idx].realY = y;
+                }
+            }
+        },
+        
+        handleEnd: function (e) {
+            var touches = this.getTouchesFromEvent(e), i, l;
+            if (this.fingers.length > 1) {
+                // If more than one finger *was* touching the screen, 
+                // find which ones are not touching anymore
+                var listed = [];
+                for (i = 0, l = touches.length; i < l; i++)
+                    listed[i] = touches[i].identifier;
+                for (i = 0, l = this.fingers.length; i < l; i++) {
+                    if (listed.indexOf(this.fingers[i].identifier) === -1) {
+                        if (this.fingers[i].entity)
+                            this.fingers[i].entity.trigger("TouchEnd");
+                        this.fingers.splice(i, 1);
+                        l--;
+                        i--;
+                    }
+                }
+            } else {
+                var pos = Crafty.DOM.translate(touches[0].clientX, touches[0].clientY),
+                  tar = e.target ? e.target : e.srcElement,
+                  closest = this.findClosestTouchEntity(pos.x, pos.y, tar);
+                if (this.fingers[0].entity) {
+                    this.fingers[0].entity.trigger("TouchEnd");
+                } else if (closest) closest.trigger("TouchEnd");
+                this.fingers.splice(0, 1);
+            }
+        },
+            
+        handleCancel: function (e) {
+            var touches = this.getTouchesFromEvent(e);
+            for (var i = 0, l = touches.length; i < l; i++) {
+                this.fingers.splice(i, 1);
+            }
+        },
+            
+        setTouch: function (touch, entity) {
+            return { identifier: touch.identifier, realX: touch.realX, realY: touch.realY, entity: entity };
+        },
+            
+        findClosestTouchEntity: function (x, y, tar) {
+            return Crafty.findClosestEntityByComponent("Touch", x, y, tar);
+        },
+           
+        fingerDownIndexById: function(idToFind) {
+            for (var i = 0, l = this.fingers.length; i < l; i++) {
+                var id = this.fingers[i].identifier;
+                
+                   if (id == idToFind) {
+                       return i;
+                   }
+                }
+            return -1;
+        },
+            
+        fingerDownIndexByEntity: function(entityToFind) {
+            for (var i = 0, l = this.fingers.length; i < l; i++) {
+                var ent = this.fingers[i].entity;
+                
+                if (ent == entityToFind) {
+                    return i;
+                }
+            }
+            return -1;
+        },
+            
+        getTouchesFromEvent: function(e){
+            var touches;
```

Comment:
I tried [switching the order here](http://jsfiddle.net/tfasjt0m/8/); defaulting to returning the changedTouches.  This fixes the particular bug I noticed. However, since you manually compare the touch event list each time to determine if a finger is lifted, it then produces false `TouchEnd" events, as `changedTouches` deliberately doesn't include previous events unless they've, well, changed.